libr/core/cmd_cmp.c

```diff
@@ -304,59 +405,189 @@ static int radare_compare(RCore *core, const ut8 *f, const ut8 *d, int len, int
 	return len - eq;
 }
 
-static void cmd_cmp_watcher(RCore *core, const char *input) {
-	char *p, *q, *r = NULL;
-	int size = 0;
-	ut64 addr = 0;
+/* Returns 0 if operation succeeded, 1 otherwise */
+static int cmd_cmp_watcher(RCore *core, const char *input) {
+	static const char *help_msg_cw[] = {
+		"Usage: cw", "[args]", "Manage compare watchers; See if and how memory changes",
+		"cw??", "", "Show more info about watchers",
+		"cw ", "addr sz cmd", "Add a compare watcher",
+		"cw", "[*qj] [addr]", "Show compare watchers (*=r2 commands, q=quiet, j=json)",
+		"cwd", " [addr]", "Delete watcher",
+		"cwr", " [addr]", "Revert watcher",
+		"cwu", " [addr]", "Update watcher",
+		NULL
+	};
+	static const char *verbose_help_cw =
+		"Watchers are used to record memory at 2 different points in time, then\n"
+		"report if and how it changed. First, create one with `cw addr sz cmd`. This\n"
+		"will record sz bytes at addr. To record the second state, use `cwu`. Now, when\n"
+		"you run `cw`, the watcher will report if the bytes changed and run the command given\n"
+		"at creation with the size and address. You may overwrite any watcher by creating\n"
+		"another at the same address. This will discard the existing watcher completely.\n"
+		"\n"
+		"When you create a watcher, the data read from memory is marked as \"new\". Updating\n"
+		"the watcher with `cwu` will mark this data as \"old\", and then read the \"new\" data.\n"
+		"`cwr` will mark the current \"old\" state as being \"new\", letting you reuse it as\n"
+		"your new base state when updating with `cwu`. Any existing \"new\" state from running\n"
+		"`cwu` previously is lost in this process. Watched memory areas may overlap with no ill\n"
+		"effects, but may have unexpected results if you update some but not others.\n"
+		"\n"
+		"Showing a watcher without updating will still run the command, but it will not report\n"
+		"changes.\n"
+		"\n"
+		"When an address is an optional argument, the command will apply to all watchers if\n"
+		"you don't pass one.\n"
+		"\n"
+		"For more details and examples, see section 4.10 of the radare2 book.\n";
+
+	ut64 addr = UT64_MAX;
+	int ret = 0;
+
 	switch (*input) {
-	case ' ':
-		p = strdup (input + 1);
-		q = strchr (p, ' ');
-		if (q) {
-			*q++ = 0;
-			addr = r_num_math (core->num, p);
-			r = strchr (q, ' ');
-			if (r) {
-				*r++ = 0;
-				size = atoi (q);
+	case ' ': { // "cw "
+		char **argv;
+		int argc;
+		ut64 size;
+
+		argv = r_str_argv (input + 1, &argc);
+		if (!argv) {
+			return 1;
+		}
+
+		if (argc == 1) { // "cw [addr]"
+			addr = r_num_get (core->num, argv[0]);
+			r_core_cmpwatch_show (core, addr, 0);
+		} else if (argc == 3) { // "cw addr sz cmd"
+			addr = r_num_get (core->num, argv[0]);
+			size = r_num_get (core->num, argv[1]);
+
+			if (size < 1) {
+				ret = 1;
+				eprintf ("Can't create a watcher with size less than 1.\n");
+				goto out_free_argv;
+			} else if (size > INT_MAX) {
```

Comment:
use r2 constants INT_MAX can be different depending on compilter. kind of ST32_MAX ?

Reply:
size is just an int, and INT_MAX should match it. its kept as an int because that's what r_io_read_at() uses and i didnt want to potentially break that